Add implementation of PML

sbastrakov opened the pull request:
I've now put my changes on top of the current dev. There are still some leftovers from the previous efforts, and, importantly, the kernel is still not changed to work in the internal area vs. the guard as originally planned. So still WIP, but hopefully just for 1-2 more days.

I'll now close the older PR #2863 .

To enable the new solver, use `YeePML` as the field solver in `fieldSolver.param`. All the PML-related parameters are in `pml.param` (default values should normally be reasonable), parameters of the old exponential damping absorber should still be specified so that the code compiles, but are not actually used. Otherwise, PML behaves exactly the same as the old absorber, in terms of absorbing area location (inside the simulation area, near the outer border) and interaction with lasers and moving window.
Known limitations:

- When PML is enabled, for each grid node on the whole domain where are 12 additional `float_X` values, so memory requirements for field data on both host and device are increased by about 2 times (particles' data is not affected, ofc). This issue will be addressed in a following PR.
- That additional values are not written to checkpoints. It is technically possible to load from a checkpoint, but field dynamics in the PML area will be changed and reflections introduced, so not worthy for meaningful simulations. This will be addressed shortly when absorbers become plugins in a following PR.
- Now field solver kernels are much more computationally intensive and use way more registers. However, in general PIC scale this should be negligible, as most time is spent on the other stages, and the field solver kernels are also interleaved with asynchronous MPI communications. Even in my tests with no particles, there is no time difference between the `Yee` + old absorber and new `YeePML`.

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/Yee/Curl.hpp:
```diff
@@ -35,9 +35,10 @@ namespace maxwellSolver
 {
 
-    template< typename  Difference >
+    template< typename T_Difference >
```
Even if this change is small please provide this change as a separate PR because it is independent to this PR.

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,414 @@
+            }
+
+            /**
+             * Propagate B values in the given area by half a time step.
```
Please put the brief description to the first line `/**`

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,414 @@
+                Thickness localThickness = globalSize;
+                for( auto exchange = 1; exchange < numExchanges; ++exchange )
+                {
+                    // Ignore directions except left, right, top, bottom, back, front
```
I would rephrase it to `Skip directions ...`

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,424 @@
+                constexpr float_X dt = DELTA_T;
+
+                // Shift indexes so that 0 is start of the local domain without guard
+                const float3_X sigmaIdx(
```
You can also write 

```C++
const float3_X sigmaIdx(
    precisionCast<float_X>(cellIdx - numGuardCells)
);
```

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,424 @@
+                    cellIdx.z() - numGuardCells.z()
+                );
+                const float3_X sigma = detail::getSigma( sigmaIdx, parameters, numLocalDomainCells );
+                const float3_X damping( 
```
we should define the dotwise `exp` function for pmacc vectors and then write 

```C++
const float3_X damping( 
     math::exp( -sigma * dt )
);
```

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,424 @@
+                    math::exp( -sigma[1] * dt ),
+                    math::exp( -sigma[2] * dt )
+                );
+                float3_X diff( dt, dt, dt );
```
can be written as `auto diff = float3_X::create( dt );`

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,424 @@
+                for ( uint32_t dim = 0; dim < 3; dim++ )
+                    if ( sigma[ dim ] )
+                        diff[ dim ] = ( 1.0_X - damping[ dim ] ) / sigma[ dim ];
+                ///std::cout << "E: cellIdx = " << cellIdx << ", damping = " << damping << ", diff = " << diff << "\n";
```
note: dead debug code

psychocoderHPC commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,424 @@
+            float_X result = 0.0;
+            if( index < numPMLCellsNegative )
+                result = ( numPMLCellsNegative - index ) / numPMLCellsNegative;
+            if( index > numLocalDomainCells - numPMLCellsPositive )
```
should be an `else if` since only one path can be valid.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/Field.tpp:
```diff
@@ -0,0 +1,135 @@
+namespace yeePML
+{
+
+Field::Field( MappingDesc cellDescription ) :
```
Thanks, fixed.

sbastrakov commented:
I made experiments on setups from #2948, including the commented out ones, to compare the absorbing performance of the exponential damping absorber and the new PML implementation. Everything seems to behave as it should. PML provides a much better absorption even with less thickness (as expected). So it should be a strictly better alternative, except additional memory for fields. 

With parameters `ALPHA_MAX_SI` and `KAPPA_MAX` (in `pml.param`) a user can switch between different versions of PML. Alpha > 0 corresponds to convolutional (CPML), zero - to uniaxial (UPML). Having Kappa > 1 means coordinate stretching in the PML area. As the literature and my experiments suggest, there is no option that is clearly superior to others. The default parameters with 8 cells of CPML and no stretching should be generally reasonable.

Attaching the results in Excel and plots separately as pictures.
[PlotsPML.xlsx](https://github.com/ComputationalRadiationPhysics/picongpu/files/3178523/PlotsPML.xlsx)
<img width="367" alt="CPML_vs_UPML_8cells_2D" src="https://user-images.githubusercontent.com/6825656/57709755-e2ca5580-766b-11e9-81b7-65e11ba4cc11.png">
<img width="367" alt="CPML_vs_UPML_8cells_3D" src="https://user-images.githubusercontent.com/6825656/57709756-e2ca5580-766b-11e9-927d-dc95d5e5d882.png">
<img width="367" alt="CPML_vs_UPML_16cells_2D" src="https://user-images.githubusercontent.com/6825656/57709757-e2ca5580-766b-11e9-8a7a-004455727e1b.png">
<img width="367" alt="CPML_vs_UPML_16cells_3D" src="https://user-images.githubusercontent.com/6825656/57709758-e2ca5580-766b-11e9-926b-cf478dfb553c.png">
<img width="423" alt="PML_vs_ExponentialDamping_2D" src="https://user-images.githubusercontent.com/6825656/57709759-e2ca5580-766b-11e9-92e9-e51c3d7e460a.png">
<img width="423" alt="PML_vs_ExponentialDamping_3D" src="https://user-images.githubusercontent.com/6825656/57709760-e362ec00-766b-11e9-966b-193de68e8950.png">

sbastrakov commented:
The code is ready for review.

PrometheusPi commented:
@psychocoderHPC since #2948 depends on this pull request, how is your plan to review this pull request?

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,431 @@
+     *
+     * Enabling this solver results in more memory being used on a device:
+     * 12 additional scalar field values per each grid cell of a local domain.
+     * Another limitation is not full persistency with checkpointing: the
```
This is a quite hard limitation. @psychocoderHPC, @ax3l and @steindev what do you think about that? Fore me this limits usability for long simulations where restarts are required. However, usually those are the ones that need better absorbers.

sbastrakov replied:
Yes, these limitations will be lifted in following PRs. The plan for that is given in the first message in this discussion.

psychocoderHPC commented:
@PrometheusPi I will review this and merge it and then we can merge #2948

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,587 @@
+         * @param cellIdx cell index including the guard, can be fractional,
+         * e.g. for halves of cells
+         * @param parameters parameters of PML in the local domain
+         * @param normalizedSigma value of normalized sigma at the cell, output
```
Why do you add `, output`? 
Same applies to two lines below.

sbastrakov replied:
Will change these to the doxygen `@param[out]`.

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,587 @@
+                         * components. Coefficients Ca, Cb as given in (7.107a,b)
+                         * are general to account for materials, in addition to
+                         * artificial PML absorbing medium. We do not have any
+                         * real material, so in (7.107a,b) have to use
```
missing `we`: ... **we** have to ...

sbastrakov replied:
Will fix.

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,587 @@
+                         * real material, so in (7.107a,b) have to use
+                         * sigma(i + 1/2, j, k) = 0 (it is another sigma,
+                         * unrelated to PML), eps(i + 1/2, j, k) = EPS0. Also,
+                         * same as the Yee scheme in PIC, adjust to use B,
```
adjust**ed**?

sbastrakov replied:
Sure.

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,587 @@
+                         * sigma(i + 1/2, j, k) = 0 (it is another sigma,
+                         * unrelated to PML), eps(i + 1/2, j, k) = EPS0. Also,
+                         * same as the Yee scheme in PIC, adjust to use B,
+                         * not H, in the right-hand side.
```
~in~ on [see here](https://www.dict.cc/englisch-deutsch/on+the+right+hand+side.html)

sbastrakov replied:
Good point.

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,587 @@
+         * @tparam T_Acc alpaka accelerator type
+         * @tparam T_Mapping mapper functor type
+         * @tparam T_Curl curl functor type
+         * @tparam T_EBox pmacc::DataBox, electric field box type
```
I think the extra comma `,`is not needed here - same is true for the line below.

sbastrakov replied:
Will fix.

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,587 @@
+                        DataSpaceOperations< simDim >::template map< SuperCellSize >( linearIdx );
+                    // grid index to process with the current thread
+                    auto const idx = blockBeginIdx + idxInSuperCell;
+                    // with the current Yee grid, a half cell shift needed here
```
missing `is`: **is** needed here

sbastrakov replied:
Thanks, will fix.

PrometheusPi commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel:
```diff
@@ -0,0 +1,587 @@
+                         * real material, so in (7.107a,b) have to use
+                         * sigma*(i + 1/2, j, k) = 0 (it is another sigma*,
+                         * unrelated to PML), mue(i + 1/2, j, k) = MUE0. Also,
+                         * same as the Yee scheme in PIC, adjust to use B,
```
adjust**ed**? (see comment before)

sbastrakov replied:
Agree.

PrometheusPi commented on include/picongpu/param/pml.param:
```diff
@@ -0,0 +1,158 @@
+namespace yeePML
+{
+
+    /* The parameters in this file are only used if field solver is YeePML.
```
Okay - perhaps: `if the field solver is set to YeePML` or  `if the field solver selected is YeePML`.

sbastrakov replied:
Good point.

PrometheusPi commented on include/picongpu/param/pml.param:
```diff
@@ -0,0 +1,158 @@
+     * for arbitrary media. Microwave and optical technology letters. 27 (5),
+     * 334-339 (2000).
+     * https://doi.org/10.1002/1098-2760(20001205)27:5%3C334::AID-MOP14%3E3.0.CO;2-A
+     * Our implementation based on a more detailed description in section 7.9 of
```
**is** based (see comment before)

PrometheusPi commented on include/picongpu/param/pml.param:
```diff
@@ -0,0 +1,158 @@
+     * values providing less reflections.
+     * 8 cells should be good enough for most simulations. There are no
+     * requirements on thickness being a multiple of the supercell size.
+     * It is only required that PML is small enough to fit near-boundary local
```
see comment/question before: This is a bit unclear to me.

sbastrakov replied:
Not sure which comment/question do you mean?

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/Parameters.hpp:
```diff
@@ -0,0 +1,122 @@
+
+        /** Max value of complex frequency shift
+         *
+         * Components correspond to directions. Normalized by eps0.
```
The complex stretching transformation is given by `x -> kappa_x + sigma_x / (alpha_x + i * omega * eps0)`, so `alpha` actually has this unit. Whether it is a good name is arguable, but this is really standard in the PML literature.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/Field.hpp:
```diff
@@ -0,0 +1,157 @@
+         *
+         * characterizing the record's unit in SI
+         * (length L, mass M, time T, electric current I,
+         *  thermodynamic temperature theta, amount of substance N,
```
Thanks, will fix.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,431 @@
+         * CFS - PML for arbitrary media. Microwave and optical technology
+         * letters. 27 (5), 334-339 (2000).
+         * https://doi.org/10.1002/1098-2760(20001205)27:5%3C334::AID-MOP14%3E3.0.CO;2-A
+         * Our implementation based on a more detailed description in section
```
Well spotted.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,431 @@
+                /* Note: here it is possible to first check if PML is enabled
+                 * in the local domain at all, and otherwise optimize by calling
+                 * the normal Yee update kernel. We do not do that, as this
+                 * would be fragile wrt future separation of PML into a plugin.
```
I meant wrt == with respect to, will write as the full expression to avoid confusion.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,431 @@
+            /** Thickness in terms of the global domain.
+             *
+             * We store only global thickness, as the local one can change
+             * during the simulation and so has to be recomputed each time step.
```
WIll fix.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,431 @@
+             * during the simulation and so has to be recomputed each time step.
+             * There are no limitations on the size, as long as it fits a single
+             * layer of external local domains (near the global simulation area
+             * boundary) on each time step. In particular, PML is not required
```
Agree this is confusing, will rephrase.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,431 @@
+                Thickness localThickness = globalSize;
+                for( uint32_t exchange = 1u; exchange < numExchanges; ++exchange )
+                {
+                    // Skip directions except left, right, top, bottom, back, front
```
Agree, will rephrase.

sbastrakov commented on include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp:
```diff
@@ -0,0 +1,431 @@
+                    if( hasNeighbour )
+                        localThickness( axis, direction ) = 0;
+
+                    // Disable PML during laser initialization
```
Yes, in terms of interactions with the lasers' `initPlaneY` and the moving window PML mimics the other absorber.

sbastrakov commented:
@PrometheusPi thanks, I've pushed a new commit addressing all your comments. To repeat about the limitations (memory and checkpoint), they will be addressed in the following PRs.
